Signbus Implementation in Tock

hsiehju opened the pull request:
## **Features** 
- Ability to act as master and write fragmented messages (longer than 255 bytes).
- Ability to act as a slave to receive fragmented messages (longer than 255 bytes).
- Able to use Hail running this kernel to request I2C bus isolation from the Signpost Controller.
- Able to use Hail running this kernel to communicate with the Signpost Controller to initialize an address and receive the correct initialization message. (tested in test_signbus_init.rs)

## **Future Plans** 
port_layer.rs
- Ability to act as slave to write messages.
- Ability to act as master to read messages.

io_layer.rs
- Send error message to dropped packet source.

protocol_layer.rs
- Add HMAC capabilities
- Add encryption/ decryption capabilities.
- Could be some userland app that handles this and talks directly to this capsule.

app_layer.rs
- Connect userland buffers to kernel buffers and add an userland callback.
- Add system calls.

test_signbus_init.rs
- Test next step in initializing a module (key_exchange, etc).

support.rs
- Move enums to specific files that use them or delete unnecessary ones.

## **General notes**
- Size of buffers are not set correctly. Running into the problem of too much RAM usage. Instead of using `&'static mut [u8]` could switch to `[u8; MAX_DATA_LENGTH]` declared on stack.

bradjc commented:
Can you add a README.md to the signbus folder that sketches out the hierarchy of files, their interfaces, and a short description of each? http://asciiflow.com/ or https://textik.com are good tools.

Something like:

```
 +--------------------------------------+
 |                                      |
 |  Another layer                       |  Does something else.
 |                                      |
 +--------------------------------------+
      interface::something::trait

+--------------------------------------+
|                                      |
|   Low Layer                          |  Handles writing on the bus.
|                                      |
+--------------------------------------+
```

I don't think we want to commit anything signpost specific (like `SignbusApiType`), so I'm trying to understand how we can make this general while still supporting signpost. Ideally I think we want to be able to run signbus on top of this, but also do other syscalls on top of it as well.

brghena commented:
You should also get rid of all the warnings that are in there right now. It looks like it's mostly unused imports (really easy to fix) and unused fields (a little harder, make sure they weren't important).

hsiehju commented:
Created a markdown called SignbusLayers.md in the folder to help explain the layers. Let me know what you think.

Added the enums  (like `SignbusApiType`) to the test_signbus_init.rs instead of the support.rs file to make it more general.

Cleared up the warnings and formatted. Travis should be passing now.

Added unimplemented!() to parts where slave write/ master read need to be implemented.

ppannuto commented on capsules/src/signbus/SignbusLayers.md:
```diff
@@ -0,0 +1,87 @@
+&nbsp;&nbsp;&nbsp;&nbsp;    -\ protocol_layer <br />
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    -\ io_layer <br />
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        -\ port_layer <br />
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        -\ I2C DRIVER <br />
```
Some unicode and a fenced code block will make this look nice in raw text and rendered form:

    ```
    USERLAND
     ↳ app_layer
      ↳ protocol_layer
       ↳ io_layer
        ↳ port_layer
         ↳ I2C DRIVER
    ```

ppannuto commented on capsules/src/signbus/app_layer.rs:
```diff
@@ -0,0 +1,127 @@
+            }
+        });
+
+        self.payload.take().map(|payload|{
```
This doesn't handle the case where the TakeCell is empty. Consider deleting line 81, and making this:

```rust
let rc = self.payload.take().map_or(ReturnCode::EBUSY, |payload| {
    self.protocol_layer.signbus_protocol_send(address, payload, len)  // Note: No ; here
});
```

https://github.com/helena-project/tock/blob/master/doc/Mutable_References.md#map-variants explains `map_or`

ppannuto commented on capsules/src/signbus/io_layer.rs:
```diff
@@ -0,0 +1,337 @@
+        self.this_device_address.set(address);
+        self.port_layer.init(address);
+
+        return ReturnCode::SUCCESS;
```
be consistent, don't need `return` here

brghena commented on boards/hail/src/main.rs:
```diff
@@ -53,12 +54,13 @@ struct Hail {
+    //si7021: &'static capsules::si7021::SI7021<'static,
+    //                                          VirtualMuxAlarm<'static, sam4l::ast::Ast<'static>>>,
+    //ninedof: &'static capsules::ninedof::NineDof<'static>,
+    //i2c_master_slave: &'static capsules::i2c_master_slave_driver::I2CMasterSlaveDriver<'static>,
```
This is a problem we've started running into and are not going to escape anytime soon. There are some pins in a platform that absolutely do not make sense to hard-code configure at compile-time rather than configure based on which applications are loaded at boot.

Ideally, we would have Signbus, at initialization time, attempt to take a lock on four GPIO pins, configuring two as GPIOs and two as I2C. If it could not take a lock (because a different application was already using them for instance) it would fail and report that problem. I'm unsure what the best way to implement this kind of thing would be though, especially in a way that is chip agnostic. @alevy Do you have any thoughts as to how best to do this?

D4 and D5 could be changed to be I2C. They are currently configured as SPI (along with D2/D3), but this was an arbitrary choice, and we could make a different arbitrary choice for now if it makes things easier.

Alternatively, we could make this a new platform (something like Hail-Signbus maybe), but that starts wandering down a path of dozens of nearly identical platforms, which feels dangerous and bad.

brghena replied:
For now, I guess my recommendation would be to bring back everything on the main I2C bus, convert D2-5 to I2C mode, use D2/3 for Mod In/Out, and use TWIMS0 for the Signbus I2C bus.

brghena commented on boards/hail/src/main.rs:
```diff
@@ -321,10 +324,10 @@ pub unsafe fn reset_handler() {
-         &sam4l::gpio::PB[11],   // D6
+        [&'static sam4l::gpio::GPIOPin; 2],
+        //[&sam4l::gpio::PB[14],   // D0
+        // &sam4l::gpio::PB[15],   // D1
```
I say we leave D0 and D1 as GPIOs, and use D2/D3 for Mod In/Out.

brghena commented on boards/hail/src/main.rs:
```diff
@@ -344,6 +347,73 @@ pub unsafe fn reset_handler() {
+       &mut capsules::port_layer::BUFFER));
+       port_layer_i2c.set_client(port_layer);
+       port_layer_virtual_alarm.set_client(port_layer);
+       */
```
Should this all be removed?

brghena commented on boards/hail/src/main.rs:
```diff
@@ -393,7 +464,25 @@ pub unsafe fn reset_handler() {
     // test_take_map_cell::test_take_map_cell();
 
-    // debug!("Initialization complete. Entering main loop");
+    // Signbus Initialization
```
This should all be commented out (but left here) like the `test_take_map_cell` code above

brghena commented on capsules/src/signbus/app_layer.rs:
```diff
@@ -0,0 +1,127 @@
+    fn packet_received(&self, data: &'static mut [u8], length: usize, error: support::Error);
+
+    // Called when an I2C master write command is complete.
+    fn packet_sent(&self, data: &'static mut [u8], error: support::Error);
```
Is this called on every I2C master write completion, or only when an entire packet has been sent (all fragments)? I think you mean the latter, and if so, update the comments for this and the others in this trait.

brghena commented on capsules/src/signbus/protocol_layer.rs:
```diff
@@ -0,0 +1,90 @@
+//! Kernel implementation of signbus_protocol_layer
+//! apps/libsignpost/signbus_protocol_layer.c -> kernel/tock/capsules/src/signbus_protocol_layer.rs
```
This is great. I'm really glad you added this mapping.

brghena commented on capsules/src/signbus/app_layer.rs:
```diff
@@ -0,0 +1,127 @@
+pub static mut BUFFER0: [u8; 256] = [0; 256];
+pub static mut BUFFER1: [u8; 256] = [0; 256];
+
+/// SignbusAppLayer to handle userland interaction and adding message type information.
```
Since this doesn't actually implement Driver (command, subscribe, etc.) this code doesn't actually directly interact with userland. Instead I'd say "to handle application messages"

brghena commented on capsules/src/signbus/SignbusLayers.md:
```diff
@@ -0,0 +1,87 @@
+
+
+## app_layer
+Userland buffers and callbacks.
```
This layer doesn't actually interact with userland (in practice if we wanted to connect this protocol to userland, there would be another layer on top of `app_layer` that would do so). Instead this handles "Application messages and callbacks", with anything above this layer being application-specific.

brghena commented on capsules/src/signbus/port_layer.rs:
```diff
@@ -0,0 +1,382 @@
+            buf[43] = 0xB3;
+
+            hil::i2c::I2CMaster::enable(self.i2c);
+            hil::i2c::I2CMaster::write(self.i2c, i2c_address, buf, 44 as u8);
```
Could you comment this all with how it "should be" when the HMAC is done in the protocol layer? Is it just removing the buf writing and changing the size to `buf.len()`?

brghena commented on capsules/src/signbus/support.rs:
```diff
@@ -0,0 +1,153 @@
+}
+
+/// Signbus packet -> [u8]
+pub fn serialize_packet(packet: Packet, data_len: usize, buf: &mut [u8]) {
```
You should do checking here to make sure you never exceed `buf.len()`

brghena commented on capsules/src/signbus/support.rs:
```diff
@@ -0,0 +1,153 @@
+}
+
+/// [u8] -> Signbus packet
+pub fn unserialize_packet(buf: &[u8]) -> Packet {
```
Check `buf.len()` here too. You shouldn't assume that the buffer is long enough to create a valid packet.

alevy commented:
At this point, do we think we'll get this to a place soon where we'll merge this into this repo? If not, perhaps this is better suited as a PR in the Signpost repo.

Does it make sense to re-brand this as an exploration in general purpose inter-IC communication?

alevy commented:
Ping @bradjc: perhaps this is better suited as a PR in the Signpost repo?

bradjc commented:
Yeah, probably. Maybe we can trick brghena into addressing his comments before we move it??

alevy commented:
Gonna close this PR for now since it hasn't been touched in a while.
